### ISSUE TITLE
Parallel playbook execution

### ISSUE THREAD

**Issue by jhawkesworth** [ansible/proposals#31]
# Proposal: Enhance Ansible to allow playbooks to be run in parallel.

_Author_: Jon Hawkesworth @jhawkesworth

_Date_: 2016/10/07
- Status: New
- Proposal type: Enhancement to core (or possibly new CLI tool)
- Targeted Release: To be decided
- Associated PR: None
- Estimated time to implement: weeks
## Motivation

Describe the reasons for this proposal.
During complex deployments, in order to keep the period of change down to the smallest possible time period, it would be very helpful if Ansible would allow (included) playbooks to be run in parallel.
### Problems

What problems exist that this proposal will solve?
- Currently I am using Jenkins to wrap up playbooks so that I can simultaneously update middleware and computer components at the same time as updating web services and front end webservers.  Wrapping each playbook in a jenkins job is a tedious manual process which I would like to replace with a top-level playbook.
## Solution proposal
- One way would be to enhance playbook includes so that a list of included playbooks causes a fork and runs all the playbooks in the list simultaneously.

So may top level playbook could look like this

```
    - include: find_release_versions.yml
    - include: synchronize_components_to_datacenter.yml
    - include: pre-deployment-checks.yml
    - include: start-deployment-notifications.yml
    - include:
         - migrate-database.yml
         - update-compute-nodes.yml
         - update-dequeuers.yml
         - update-web-services.yml
         - update-file-transfer.yml
         - update-frontend-webservers.yml
    - include: end-deployment-notifications.yml
    - include: generate-change-report.yml

```

So everything from migrate-database.yml to update-frontend-webservers.yml would be executed simultaneously and end-deployment-notifications would only start once all the listed items in the previous include had completed.

Rather than overloading 'include' in this way, another keyword might be preferable.  Ideas I have had so far for alternative keyword are 'parallel', 'together', 'fork', or 'join'.

Another implementation might be a separate CLI tool, distinct from ansible-playbook which ran playbooks listed in its own file.  In keeping with the sports name I came up with the name 'league' for the command and 'league-table' as a name for the file format.  On balance I like this idea less as it seems like much more work.  However, you might want to add the ability to specify forks or other playbook command line options (--check perhaps) for each of the parallel playbooks.

Keeping the change within ansible-playbook would make integration with tower and other tools easier.

Here's what a playbook might look like with parallel and some options passed to individual playbooks.

```
    - include: find_release_versions.yml
    - include: synchronize_components_to_datacenter.yml
    - include: pre-deployment-checks.yml
    - include: start-deployment-notifications.yml
    - parallel:
         - { playbook: 'migrate-database.yml', forks: 1, check: True}
         - { playbook: 'migrate-database.yml', forks: 1 }
         - update-compute-nodes.yml
         - update-dequeuers.yml
         - update-web-services.yml
         - update-file-transfer.yml
         - { playbook: 'update-frontend-webservers.yml', forks: 20 }
    - include: end-deployment-notifications.yml
    - include: generate-change-report.yml

```
## Dependencies

I can't think of any dependencies that aren't already in use in ansible-playbook
## Testing

If passing playbook options is included in this functionality, there would be significant testing.  
Possibly useful to enable testing, although this seems to be handled in Shipable at the time of writing.
## Documentation

I can imagine that this feature could possibly cause ansible controllers to run out of resources as it might cause many forks to be generated, so it would need documenting to advise users not to go crazy with it.

**Comment by MichaelBaydoun** [comment]
This would be very useful to us.  One use case, we get to a point in our playbook when we need to spin up 4 different aws instances, configure them, and create ami's from the result.  Doing that in Parallel would save a lot of time.

**Comment by hyperized** [comment]
This would save a lot of time by rolling out changes in common roles across our hundreds of servers :).

**Comment by alikins** [comment]
For the parallel invocation case

``` yaml
- parallel:
         - { playbook: 'migrate-database.yml', forks: 1, check: True}
         - { playbook: 'migrate-database.yml', forks: 1 }
         - update-compute-nodes.yml
         - update-dequeuers.yml
         - update-web-services.yml
         - update-file-transfer.yml
         - { playbook: 'update-frontend-webservers.yml', forks: 20 }
```

What is the general expectation for how the parallel playbooks interact with each other?

-   if update-compute-nodes.yml adds hosts to the inventory, how would the other playbooks react? Do they see it in inventory immediately? After some stage? never? 

- In this example, is each playbook targeting a different set of hosts?
  - if they target the same hosts, is there any ordering implied? 
  - If a tasks ends up conflicting with another (say, one has 'file: name=/somefile, state=absent' and another has 'file: name=/somefile, state=present') how would/should/could that be resolved?

- would handlers be expected to work across playbooks?

Granted, those scenarios could be consider as errors on the playbook authors part, so may just be 'dont do that then'.

**Comment by bcoca** [comment]
Funny enough, I show how several ways to do this, without altering ansible http://www.slideshare.net/bcoca/more-tips-n-tricks (starts at slide 20)

**Comment by jhawkesworth** [comment]
@alikins Thanks, all good points.
My expectation would be that inventory would be determined as each playbook begins execution.  So in the example above you would want to have an 
`- include: update-inventory.yml `
before the 
`parallel: `
update-compute-nodes.yml would probably be too late.

Yes, my assumption is that each of the playbooks listed in the parallel: list would be targeting different hosts (this is based on my actual use case).  Targeting the same hosts from different playbooks sounds like a 'don't do that unless it really makes sense'.

I would not want any ordering implied - so a yaml list is probably a bit of a confusing way of notating it.

I would want handlers to have a scope that only exists within each invoked playbook (which I suspect is not the case at the moment) so that could be a deal breaker.  I haven't given this any significant thought as I don't use handlers in any of my roles, so I would welcome comment from anyone who uses handlers and playbook includes.

Another assumption in the above is all playbooks are targeting the same inventory which seems potentially inflexible (especially if inventory and handlers were scoped to the playbook level).

@bcoca - Thanks for the slides (wish I could have been there).  I will experiment and see if your 'trying really hard to avoid shell scripts v3' method meets my needs.

**Comment by bcoca** [comment]
So to extend my view on this, there are plenty of external systems that allow you to do this w/o building it into ansible (my slides above show some easy/crude examples).

To add this feature we cannot do this with the current main object (play) it would have to be a 'super playbook' that both contains the inventory and list of playbooks and will require all new DSL to describe this object and deal with it having playbooks (or plays?), then there is the handling of output and error conditions ... and that is just from a first thought, i'm sure plenty of other issues will arrise.  

To add this supra layer and complexity to Ansible itself does not seem warranted when there are a plethora of execution systems that can invoke Ansible in a parallel fashion, starting by Tower (shameless plug for the thing that pays my salary) but not limited to it as xargs, parallels, condor, jenkins, cron, at, celery, hadoop ... anything that can execute a command line program in parallel with different options exist already.  

I don't see this as something Ansible should tackle and way outside it's purview and adds extra layers of complexity that are really not needed.

**Comment by hyperized** [comment]
@bcoca Having experimented today with the examples from the slide above and I came across two immediate issues:

1. Because of that use, ad-hoc groups cannot be used.
2. Because of the invoking of the playbooks you introduce new requirement for paths and (heaven forbid) things like sudo prompts.

I also found that it's also tricky to deal with feedback, for example: I will still require output from the scripts to be captured, so I had to opt back to a poll time, breaking the parallelism.

Those two use-cases above were preventing me to allow for proper parallelism with complicated tasks such as coordinated cluster reboots within a small maintenance window. The solution currently is to simply schedule multiple task in an external task scheduler. But this proposal would perhaps bring a more native solution to the table.

**Comment by bcoca** [comment]
@hyperized my slides are a hint, not a solution, specially the ones that use ansible to shell out to run ansible, that is a 'negative example'. 

As mentioned above, there are plenty of other external schedulers that already can do this much better and I really see no need of a native solution.

**Comment by jhawkesworth** [comment]
I still think a native-to-ansible way of running playbooks in parallel which is declarative would enhance ansible's powers as a tool for orchestrating change.  When I say orchestrating, I'm talking in the 'conducting the orchestra' sense.  Like all good orchestral conductors, I think ansible should be able to get the members of the orchestra to play different stuff at the same time. :-)

However, I'm going to close this proposal because:

The time to do it is not now.
I have no time to create something.

**Comment by Ehekatl** [comment]
This would be an really useful feature for us

in our case we have a playbook include many other playbooks run on different host group, parallel execution would make this a lot more efficient

**Comment by jhawkesworth** [comment]
I wound up using a bash script that calls the playbooks using gnu parallels.  This works ok, parallels doesn't return any output until the playbook has run through or failed, so we tend to wind up tailing the ansible log file or syslogs

**Comment by dagwieers** [comment]
Thinking about this I would tend to do `include:` using async.

The problem obviously is that include doesn't really return status information of the processed 'tasks'. So maybe that's what is missing currently. Make include/import report back **failed**/**changed** and allow async.

Maybe this is already possible, I haven't tested this, but IMO that would be the most in line with how Ansible currently works, and would offer a lot of flexibility.
